File: Kadanov/Kadan.go
package main

import (
    "fmt"
    "time"
	"bufio"
	"os"
	"strconv"
	"io"
	"math"
)

func kadane(arr []int) int {
  maxEndingHere := 0
  maxSoFar := math.MinInt32

  for i := 0; i < len(arr); i++ {
    maxEndingHere = maxEndingHere + arr[i]
    if maxEndingHere < 0 {
      maxEndingHere = 0
    }
    if maxSoFar < maxEndingHere {
      maxSoFar = maxEndingHere
    }
  }

  return maxSoFar
}

func readNumbersFromFile(filename string) ([]int) {
	file, _ := os.Open(filename)

	defer file.Close()

	var numbers []int

	reader := bufio.NewReader(file)
	for {
		line, _ , err:= reader.ReadLine()
		if err == io.EOF {
			break
		}
		num, _ := strconv.Atoi(string(line))
		numbers = append(numbers,num)
	}
	return numbers
}

func main() {
	
  seznam := readNumbersFromFile("numbers.txt")
  startTime := time.Now()
  kadane := kadane(seznam)
  endTime := time.Now()
  timeTaken := endTime.Sub(startTime).Milliseconds()
  fmt.Println(kadane)
  fmt.Println("Time taken: %d milliseconds\n", timeTaken)
}

File: Kadanov/Kadan.py
import time

def kadane(arr):
  max_ending_here = 0
  max_so_far = -float("inf")

  for i in range(len(arr)):
    max_ending_here += arr[i]
    if max_ending_here < 0:
      max_ending_here = 0
    if max_so_far < max_ending_here:
      max_so_far = max_ending_here

  return max_so_far

def read_integers_from_file(file_path):
    try:
        seznam = []
        with open(file_path, 'r') as file:
            while True:
                number = file.readline()
                if not number:
                    break
                seznam += [int(number)]
            return seznam
    except FileNotFoundError:
        return []
        
seznam = read_integers_from_file("numbers.txt")
startTime = time.process_time()
maks = kadane(seznam)
endTime = time.process_time()
print(maks)
timeTaken = (endTime - startTime) * 1000.0
print(f"Execution time: {timeTaken:.3f} milliseconds")
    